[QoI] Properly diagnose closure parameter distructuring after SE-0110

xedin opened the pull request:
Swift 3 supported limited argument destructuring when it comes to
declaring (trailing) closures. Such behavior has been changed by
SE-0110. This patch aims to provide better error message as well
as fix-it (if structure of the expected and actual arguments matches)
to make the migration easier and disambiguate some of the common
mistakes.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-4738](https://bugs.swift.org/browse/SR-4738).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

slavapestov commented:
This needs tests. Also the parameter destructuring logic looks familiar, I'll see if I can find the similar code tomorrow.

xedin commented:
@jrose-apple This still needs some polish but it's ready for preliminary review, I haven't changed/added tests yet, but after running with `-T` it looks like this fixes at least 2 crashers and improves multiple diagnostics.

Pitfalls are mostly related to situations like `{ (x: Double, y: Double) in x + y }` - this changes are going to generate incorrect fix-it `let (x: Double, y: Double) = arg` but I think I got that covered and going to generate `let x: Double = arg.0` and `let y: Double = arg.1` instead. Also if there were no parenthesis e.g. `{ x, y in x + y }` fix-it is going to be incorrect as well `let x, y = arg`.

xedin commented:
@slavapestov You are too fast, I was just writing comment to @jrose-apple about that :)

slavapestov commented:
Also you might need to check for Swift version 3. We're not hanging the old behavior with SE-0110

xedin commented:
Yeah, I'm aware of that...

xedin commented:
One of the examples of improved diagnostics:
```
(swift) let t = (1, (2, 3))
// t : (Int, (Int, Int)) = (1, (2, 3))
(swift) [t].map { (x, (y, z)) -> Int in x + y + z }
<REPL Input>:1:11: error: closure argument list does not support tuple destructuring
[t].map { (x, (y, z)) -> Int in x + y + z }
          ^~~~~~~~~~~
          (arg)                 let (x, (y, z)) = arg;
```

xedin commented:
@jrose-apple @rudkx Also made it makes sense instead of suggesting replacing existing arguments with (arg), suggest to remove parameters all together and use `let ... = $0` instead?

xedin commented:
Ah, I guess it doesn't work if there is an explicit result type specified, no never mind...

jrose-apple commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,39 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+
+    diagnose(params->getStartLoc(), diag::anon_closure_arg_tuple_destructuring)
+        .fixItReplace(params->getSourceRange(), "(arg)")
+        .fixItInsertAfter(Lexer::getLocForEndOfToken(SourceMgr, inLoc),
```
Instead of "after the `in`", this should be "before `Tok`" (the first token of the body, right?). I'm not sure how to preserve the indentation for the actual first line, though.

jrose-apple commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,39 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+      OS << '\n';
+
+    OS << "let " << SourceMgr.extractText(paramRange) << " = arg"
+       << (Tok.isAtStartOfLine() ? '\n' : ';');
```
`"; "`, please.

jrose-apple commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,39 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+
+    auto paramRange = CharSourceRange(
+        SourceMgr, params->getStartLoc(),
+        Lexer::getLocForEndOfToken(SourceMgr, params->getEndLoc()));
```
Sorry I've been so confusing with this! At Parse time, we don't actually want to replace the entire argument list, just the part that's being destructured. So `(x, (y, z))` should become `(x, arg2)`.

xedin replied:
Ah ok, for for example (x, (y, z), (a, b)) is going to result in (x, arg1, arg2) and 2 "let" statements - "let (y, z) = arg1" and "let (a, b) = arg2", WDYT?

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6454,6 +6454,50 @@ visitRebindSelfInConstructorExpr(RebindSelfInConstructorExpr *E) {
 }
 
+static bool matchArgStructure(Type lhs, Type rhs,
+                              std::function<bool(Type, Type)> comparator) {
```
LLVM has a slightly cheaper container for non-escaping callbacks, llvm::function_ref.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6454,6 +6454,50 @@ visitRebindSelfInConstructorExpr(RebindSelfInConstructorExpr *E) {
+  flattenType(lhs, componentsA);
+  flattenType(rhs, componentsB);
+
+  // There is something structuraly invalid with arguments.
```
Typo: "structurally" (here and below)

xedin replied:
Sorry and thanks!

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6454,6 +6454,50 @@ visitRebindSelfInConstructorExpr(RebindSelfInConstructorExpr *E) {
+      auto tupleA = typeA->getAs<TupleType>();
+      auto tupleB = typeB->getAs<TupleType>();
+      matches = tupleB && tupleA->getNumElements() == tupleB->getNumElements();
+    } break;
```
Style nitpick: we tend to just put the `break` inside the braces.

xedin replied:
Will do!

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6554,75 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+
+            // Can't produce fix-it, because inferred parameter
+            // and actual argument type structure are not the same.
+            if (!matchArgStructure(actualArgType, argTupleTy, typeComparator))
```
I don't understand this test. It falls down in the presence of things like Optional or base classes of generic types, for which `visit` will return a different number of types. What are you actually protecting against?

jrose-apple replied:
I'm pretty sure you can just drop this. Fix-its on notes are allowed to be wrong sometimes, but even then I can't think of a situation where it would be.

xedin replied:
Ok, sounds good I wasn't sure it everything has to be exactly correct or not, I'll just remove matchArgStructure completely.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6554,75 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+              if (auto fixedType = CS->getFixedType(typeVar))
+                return fixedType && fixedType->isEqual(rhs);
+
+              // type variabls in this case represent types of
```
Typo: "Type variables"

xedin replied:
Sorry and thanks!

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6554,75 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+            // in the suggested 'let' to keep the structure of the code in
+            // tact, otherwise just use ';' to keep everything on the same
+            // line.
+            if (!CE->hasSingleExpressionBody())
```
`hasSingleExpressionBody` is not the same condition is "is written on multiple lines". I think it's worth really looking for a newline to decide whether to insert newlines.

On the other hand, `hasSingleExpressionBody` *does* indicate that the body was previously available for type inference and now it won't be, and so we should include the parameter type in the fix-it.

jrose-apple reviewed (commented):
I think this is looking pretty good! But there are a lot of pieces that aren't quite lining up yet, and of course it does need the test cases. Please let me know if anything I've said here is unclear.

jrose-apple commented:
By the way, for anyone following along: originally Pavel wanted to write the whole thing in Sema, where we have enough type information to jump directly to the correct solution. However, I was uneasy about letting something like `(x, (y, z)) in …` even survive to Sema, even though it previously got parsed as `(x, _: (y, z)) in …`. So now there are two similar fix-its that address different problems.

slavapestov commented:
Why not let it survive to Sema?

xedin commented:
I've addressed all of your comments @jrose-apple and added tests, couple of things are still ongoing:
1. I'll trying to figure out how to validate fix-its and not just errors;
2. Will try to tackle the situation from https://bugs.swift.org/browse/SR-2135.

xedin commented:
@swift-ci please test

rudkx commented:
@xedin I think it's less important to handle things that weren't legal Swift 3 code, so you can always tackle https://bugs.swift.org/browse/SR-2135 as a follow-on commit.

rudkx commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,39 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+  // If this was a closure declaration (maybe even trailing)
+  // it's imperative to detect any parameter list destructuring
+  // early on, see SE-0110 for more details.
+  for (unsigned i = 0, e = params->size(); i != e; ++i) {
```
Minor nit: This makes it sound like SE-0110 has details about why it's imperative to detect parameter list destructuring. You might want to turn that comma into a period and then add a line explaining why it's important to do this here.

xedin replied:
Will do, thanks!

rudkx commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -266,10 +266,18 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
-        // Unnamed parameters must be written as "_: Type".
+        // If this is a closure declaration, what is going
+        // on is most likely argument destructuring, we are going
+        // to diagnose that after all of the paramters are parsed.
```
Nit: typo - parameters not paramters

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,69 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+        // Let's see if inferred argument is actually a tuple inside of Paren.
+        if (auto *argTupleTy = inferredArgType->getAs<TupleType>()) {
+          // Looks like the number of closure parameters matches number
+          // of inferred arguments, which means we can warn about SE-0110
```
I would avoid referencing the SE proposal like this because it results in an extra step to go figure out what you're talking about. You could just say "...we can emit an error about an attempt to make use of tuple splat or tuple destructuring..."

xedin replied:
Will do!

rudkx commented on test/Constraints/tuple_arguments.swift:
```diff
@@ -1336,26 +1336,26 @@ do {
-  s.takesClosureTuple({ _ = $0; _ = $1 }) // expected-error {{cannot convert value of type '(_, _) -> ()' to expected argument type '((Double, Double)) -> ()'}}
-  s.takesClosureTuple({ (x, y) in }) // expected-error {{cannot convert value of type '(_, _) -> ()' to expected argument type '((Double, Double)) -> ()'}}
-  s.takesClosureTuple({ (x: Double, y:Double) in }) // expected-error {{cannot convert value of type '(Double, Double) -> ()' to expected argument type '((Double, Double)) -> ()'}}
+  s.takesClosureTuple({ _ = $0; _ = $1 }) // expected-error {{closure parameter list '(Double, Double)' does not support tuple destructuring with implicit parameters}}
```
This seems a bit odd since the parameter here is a single tuple with (Double, Double), but printing it this way after the word 'list' makes it seem like there are in fact multiple arguments here.

The same applies to other errors below.

Is it possible to make this something like "closure tuple parameter of type '(Double, Double)' cannot be destructured using implicit parameters"?

xedin replied:
Yes, absolutely! I will change that since i already detect that case, just need to adjust the message.

rudkx replied:
@jrose-apple Might have further opinions on wordsmithing here, too.

jrose-apple replied:
Your phrasing looks good to me.

xedin commented:
@rudkx regarding https://bugs.swift.org/browse/SR-2135 I will make it a follow up then, since changes are going to be in parser.

xedin commented:
@jrose-apple @rudkx I've fixed the message and nits that you've mentioned, also added proper fix-its for situations like `{ x, y in ... }` and `{ (x: Int, y: Int) in ... }`. Going to try to figure out how to validate
fix-its (which I still no clue how to do) next and that should be it.

rudkx commented:
@xedin You can validate fixits from within the test cases. I'll be the first to admit that I pretty much had to reverse-engineer the syntax for this, though, so I am not of much help in doing this. See e.g. 4f5b80dd03c59aa0856832191c29a53cc502d5fd.

xedin commented:
@rudkx Yes! I looked at the some other tests in diagnostics.swift and remembered! :) All of the fix-its added now to new and changed tests.

xedin commented:
@swift-ci please smoke test

jrose-apple commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,41 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+      OS << '\n';
+
+    OS << "let " << SourceMgr.extractText(paramRange) << " = " << argName
+       << (Tok.isAtStartOfLine() ? "\n" : "; ");
```
It would be nice to also properly indent the *next* line as well. It looks like Lexer::getIndentationForLine can do that.

jrose-apple commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,41 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+  // possible and give a proper fix-it. See SE-0110 for more details.
+  for (unsigned i = 0, e = params->size(); i != e; ++i) {
+    auto *param = params->get(i);
+    if (!param->isInvalid())
```
Since there might be other ways a parameter could be invalid, it's probably best to check that it's actually missing a name and that the type is a TupleTypeRepr.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,104 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+              isMultiLineClosure = bodyLine > inLine;
+            }
+
+            std::string parameter;
```
Nitpick: SmallString?

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,104 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+
+            // If at least one of the parameters has explicitly
+            // declared type let's turn fix-it into N 'let' expressions
+            // to be able to specify it.
```
I'm still not sure why this is necessary. `let` supports destructuring and type specification at the same time.

```swift
let (x: Int, y: Double) = (0, 0)
```

The real problem with explicitly-typed parameters is that they may have affected the closure's type (see below).

xedin replied:
That's what I thought as well but I actually tried it and that didn't seem to work at all:

```
(swift)  let (x: Int, y: Int) = (0, 0)                                                                                                                                                                                                                                          <REPL Input>:1:18: error: invalid redeclaration of 'Int'
 let (x: Int, y: Int) = (0, 0)
                 ^
<REPL Input>:1:10: note: 'Int' previously declared here
 let (x: Int, y: Int) = (0, 0)
         ^
```

rudkx replied:
That's correct, the proper way to write that is:
``
let (x, y): (Int, Double) = (0,0)
``

Writing:
``
let (x: Int, y: Double) = (0, 0)
``
creates a pattern_tuple with element names x and y.

xedin replied:
I think it would be best to leave it as let statements then, WDYT?

jrose-apple replied:
I tried it and it works. What am I missing?

jrose-apple replied:
Ah, I'm sorry, I had completely misinterpreted what my code was doing. I still don't think we need this, though; we need the explicit type on the parameters regardless, and then it's unnecessary to have it on the destructuring.

rudkx replied:
@jrose-apple Things are not as they seem!

``
let (x: Int, y: Double) = (0, 0)
let z = x + y // error: use of unresolved identifier 'x', 'y'
let z = Int + Double // works!
``

CodaFi replied:
What is the right-hand clause of this production even used for?

```
tuple-pattern-element → pattern | identifier:pattern”
```

rudkx replied:
@CodaFi I don't know all the details but I looked into it at some point and Jordan reminded me this morning that it was related to tuple shuffles:
```
let (x: y, z: w) = (z: 2, x: 1)
print(y, w) // prints 1, 2
```

I believe for switch/case we do not actually allow the shuffle. I am not sure there is any value in supporting the existing shuffle behavior and it is a maintenance burden so I was hoping to someday write up a proposal to deprecate and then remove it.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,104 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+                 << (isMultiLineClosure ? "\n" : "; ");
+            }
+
+            diag.fixItReplace(params->getSourceRange(), "arg")
```
If we're starting with a single-expression closure, this probably should include the type (`(arg: (Int, Double))`). It also should include the return type if hasn't already been specified and isn't Void.

If we're starting with a multi-statement closure, the return type is fine, but the arguments should still have a type if any of the parameters-as-written had a type.

akyrtzi commented:
Is it possible (as a follow-up) that we emit a warning in Swift 3 that the code is depending on behavior incompatible with Swift 4, and have a fixit on the warning to make the code compatible with both Swift 3 & 4 ?

xedin commented:
@akyrtzi I will take a look at that?

xedin commented:
@rudkx @jrose-apple addressed comments about explicit types added to the new argument, return prepended to the single statement closure with expected return type and indentation for multi-line closures.

xedin commented:
@swift-ci please smoke test

rudkx commented on test/decl/func/functions.swift:
```diff
@@ -104,7 +104,7 @@ _ = nullaryClosure(0)
 // expected-error@+1{{unnamed parameters must be written}} {{27-27=_: }}
 func destructureArgument( (result: Int, error: Bool) ) -> Int {
-  return result  // expected-error {{use of unresolved identifier 'result'}}
+  return result
```
Does this somehow no longer emit an error?

xedin replied:
This is the result of properly marking tuple parameter is "invalid" which wasn't happening before, it still is going to emit an error about "_:" but it's going to avoid considering "result" as an identifier.

rudkx reviewed (approved):
LGTM.

xedin commented:
@jrose-apple I forgot to add check for name + TupleTypeRepr you mentioned, added now, please take a look when you'll have time.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,125 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+
+            // If this is multi-line closure we'd have to insert new lines
+            // in the suggested 'let' to keep the structure of the code in
+            // tact, otherwise just use ';' to keep everything on the same
```
Ah, "intact" is one word.

jrose-apple reviewed (commented):
I *still* don't like the `arg.1`, `arg.2` thing. Why choose that over `let (x, y) = arg`?

xedin commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,54 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+    OS << "let " << SourceMgr.extractText(paramRange) << " = " << argName
+       << (isMultiLine ? "\n" + indent : "; ");
+
+    if (!isMultiLine && explicitResultType)
```
Alright, going to remove it from both Sema and Parser...

xedin commented:
@jrose-apple it's going to emit `arg1` etc. only in case some of the arguments have types specified e.g. this situation `(x: Int, y: Int)` which if then provides a proper fix-it `let x: Int = arg.0` and `let y: Int = arg.1` instead of incorrect one `let (x: Int, y: Int) = arg` because we don't re-create of parse the type structure it's easier to split into multiple 'let' statements...

xedin commented:
@jrose-apple Ok, removed 'return' and fixed typo

xedin commented:
@swift-ci please smoke test

xedin commented:
@jrose-apple Alright, I've changed fix-it for types like `(x: Int, y: Int)` to be `(arg: (x: Int, y: Int))` and `let (x, y) = arg;`

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

jrose-apple commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2399,6 +2399,51 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
+    if (isMultiLine)
+      OS << '\n' << indent;
+
+    OS << "let " << SourceMgr.extractText(paramRange) << " = " << argName
```
This piece here still won't do the right thing for `(x: Int, (y: Int, z: Int))`, but it'd be a bit trickier to get that right.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,100 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+
+            // If at least one of the parameters has explicitly
+            // declared type let's turn fix-it into N 'let' expressions
+            // to be able to specify it.
```
Comment is now out of date.

xedin replied:
I don't believe it is - because in situation like map { $0 + $1 } we are not going to provide fix-it just say that this is tuple restructuring with implicit parameters and it's not supported...

jrose-apple replied:
That's not what the comment says, though. It says it's going to use N 'let' expressions.

xedin replied:
Ah, I don't know how I missed that, will fix it, thanks!

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,100 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+            bool explicitTypes = false;
+            for (auto *param : params->getArray()) {
+              if (!param->getType()->is<TypeVariableType>()) {
+                explicitTypes = true;
```
It's better to test this by looking at the TypeLoc for the parameter.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,100 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+
+            diag.fixItReplace(params->getSourceRange(), nameOS.str())
+                .fixItInsertAfter(
+                    Lexer::getLocForEndOfToken(sourceMgr, CE->getInLoc()),
```
Is it possible to get the start loc of the first thing in the closure, rather than the location of the `in`? Like in the parser that would produce better results.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6508,7 +6510,100 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+              nameOS << "(arg: " << argTupleTy->getString() << ")";
+            } else {
+              nameOS << "(arg)";
+            }
```
We may also need to insert an explicit return type—if it was previously a single-expression closure (`hasSingleExpressionBody`), and there's no existing return type, *and* the inferred return type isn't Void, it's probably better to not make the user encounter another error.

jrose-apple reviewed (commented):
Yet more comments, but they're all smaller by now. If you want to merge this and then continue improving it after that's fine with me!

xedin committed:
[QoI] Properly diagnose closure parameter distructuring after SE-0110

Swift 3 supported limited argument destructuring when it comes to
declaring (trailing) closures. Such behavior has been changed by
SE-0110. This patch aims to provide better error message as well
as fix-it (if structure of the expected and actual arguments matches)
to make the migration easier and disambiguate some of the common
mistakes.

Resolves: SR-4738, SR-4745, rdar://problem/31892961.

xedin commented:
@jrose-apple Addressed all of the problems you mentioned, thanks again for the review!

xedin commented:
@swift-ci please smoke test

xedin commented:
@jrose-apple Looks like CI is green, please let me know if you are ok with me merging with in the current state and maybe continuing in the separate PR.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -6608,7 +6610,144 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+
+              if (auto stmt = bodyStmts.front().get<Stmt *>()) {
+                // If the body is a single expression with implicit return.
+                if (stmt->getKind() == StmtKind::Return && stmt->isImplicit()) {
```
`isa<ReturnStmt>`

jrose-apple commented:
Yep, you're welcome to merge. Thanks for putting so much effort into this!

xedin commented:
@jrose-apple I've fixed the comment and switched to `isa<ReturnStmt>`, thanks for the review!

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge
